disco/extensions/upgrade_simulation/upgrades/voltage_upgrade_functions.py

```diff
@@ -1395,7 +1453,8 @@ def perform_clustering(num_clusters, square_distance_array, buses_with_violation
 
     """
     clusters_dict = {}
-    model = AgglomerativeClustering(n_clusters=num_clusters, affinity='euclidean', linkage='ward')
+    # model = AgglomerativeClustering(n_clusters=num_clusters, affinity='euclidean', linkage='ward')
+    model = AgglomerativeClustering(n_clusters=num_clusters, affinity='precomputed', linkage='average')

```

Comment:
Is this what fixed the issue? Do you need to keep the old line?

Reply:
No actually, this isn't what fixed the issue. By adding more interim plots, from observation, I added some more settings sweeps after LTC placement, which further reduced violations, before entering the new regulator placement module. 

Regarding the change in clustering, since we are using the distance matrix: this change makes sense. However, we could still get different groups when run multiple times.

As for the original error, I wasn't able to reproduce it again as such. I did add a new test config where multiple regulators are needed. It produced slightly varying results, but no error.

Reply:
Also, is it fine to leave the old line in there, it'll be nice to have it in there for reference, if we want to try anything else later.